lib/Dialect/FIRRTL/OpFolds.cpp

```diff
@@ -50,6 +50,27 @@ OpFoldResult ConstantOp::fold(ArrayRef<Attribute> operands) {
   return valueAttr();
 }
 
+OpFoldResult DivPrimOp::fold(ArrayRef<Attribute> operands) {
+  APInt value;
+
+  /// div(x, x) -> 1
+  ///
+  /// Division by zero is undefined in the FIRRTL specification. This
+  /// fold exploits that fact to optimize self division to one.
+  if (lhs() == rhs() && rhs().getType() == getType()) {
+    auto width = getType().cast<IntType>().getWidthOrSentinel();
+    if (width == -1)
+      width = 2;
+    return IntegerAttr::get(IntegerType::get(getContext(), width), 1);
+  }
+
+  /// div(x, 1) -> x
+  if (matchPattern(rhs(), m_FConstant(value)) && value.isOneValue())
+    return lhs();
```

Comment:
This should probably check to see if lhs().getType() == getType().  I think it can differ for signed divides.

Reply:
I had to think about this, but I agree with this. Signed divides are width expanding and that should add a sign extension op (which moves that outside of what is possible in a fold). I've updated this in the latest `!squash` commit.